File: zlcore/src/main/java/com/zaitunlabs/zlcore/models/BookmarkModel.java
package com.zaitunlabs.zlcore.models;

import android.provider.BaseColumns;

import com.activeandroid.Model;
import com.activeandroid.annotation.Column;
import com.activeandroid.annotation.Table;
import com.activeandroid.query.Select;

import java.io.Serializable;
import java.util.Calendar;
import java.util.Date;
import java.util.List;

/**
 * Created by ahsai on 3/18/2018.
 */

@Table(name = "Bookmarks", id = BaseColumns._ID)
public class BookmarkModel extends Model implements Serializable {

    @Column(name = "title")
    private String title;

    @Column(name = "desc")
    private String desc;

    @Column(name = "link")
    private String link;


    public BookmarkModel(){
        super();
    }
    public BookmarkModel(String title, String desc, String link) {
        super();
        this.title = title;
        this.desc = desc;
        this.link = link;
    }

    @Column(name = "timestamp", index = true)
    public Date timestamp;

    public void saveWithTimeStamp(){
        timestamp = Calendar.getInstance().getTime();
        save();
    }

    public String getTitle() {
        return title;
    }

    public void setTitle(String title) {
        this.title = title;
    }

    public String getDesc() {
        return desc;
    }

    public void setDesc(String desc) {
        this.desc = desc;
    }

    public String getLink() {
        return link;
    }

    public void setLink(String link) {
        this.link = link;
    }


    public boolean hasSaved(){
        return new Select().from(BookmarkModel.class).where("link = '"+link+"'").count() > 0;
    }

    public static BookmarkModel findBookmark(String link){
        return new Select().from(BookmarkModel.class).where("link = '"+link+"'").executeSingle();
    }

    public static boolean bookmark(String title, String desc, String link){
        BookmarkModel existingModel = findBookmark(link);
        if(existingModel == null){
            BookmarkModel newBoomark = new BookmarkModel(title,desc,link);
            newBoomark.saveWithTimeStamp();
            return true;
        }
        return false;
    }

    public static boolean unBookmark(String title, String desc, String link){
        BookmarkModel existingModel = findBookmark(link);
        if(existingModel != null){
            existingModel.delete();
            return true;
        }
        return false;
    }


    public static List<BookmarkModel> getAllBookmarkList(){
        return new Select().from(BookmarkModel.class).execute();
    }

    @Override
    public String toString() {
        return "BookmarkModel{" +
                "title='" + title + '\'' +
                ", desc='" + desc + '\'' +
                ", link='" + link + '\'' +
                '}';
    }
}

File: zlcore/src/main/java/com/zaitunlabs/zlcore/models/BookmarkModel.kt
package com.zaitunlabs.zlcore.models

import android.provider.BaseColumns

import com.activeandroid.Model
import com.activeandroid.annotation.Column
import com.activeandroid.annotation.Table
import com.activeandroid.query.Select

import java.io.Serializable
import java.util.Calendar
import java.util.Date

/**
 * Created by ahsai on 3/18/2018.
 */

@Table(name = "Bookmarks", id = BaseColumns._ID)
class BookmarkModel : Model, Serializable {

    @Column(name = "title")
    var title: String? = null

    @Column(name = "desc")
    var desc: String? = null

    @Column(name = "link")
    var link: String? = null

    @Column(name = "timestamp", index = true)
    var timestamp: Date


    constructor() : super() {}
    constructor(title: String, desc: String, link: String) : super() {
        this.title = title
        this.desc = desc
        this.link = link
    }

    fun saveWithTimeStamp() {
        timestamp = Calendar.getInstance().time
        save()
    }


    fun hasSaved(): Boolean {
        return Select().from(BookmarkModel::class.java).where("link = '$link'").count() > 0
    }

    override fun toString(): String {
        return "BookmarkModel{" +
                "title='" + title + '\''.toString() +
                ", desc='" + desc + '\''.toString() +
                ", link='" + link + '\''.toString() +
                '}'.toString()
    }

    companion object {

        fun findBookmark(link: String): BookmarkModel? {
            return Select().from(BookmarkModel::class.java).where("link = '$link'").executeSingle()
        }

        fun bookmark(title: String, desc: String, link: String): Boolean {
            val existingModel = findBookmark(link)
            if (existingModel == null) {
                val newBoomark = BookmarkModel(title, desc, link)
                newBoomark.saveWithTimeStamp()
                return true
            }
            return false
        }

        fun unBookmark(title: String, desc: String, link: String): Boolean {
            val existingModel = findBookmark(link)
            if (existingModel != null) {
                existingModel.delete()
                return true
            }
            return false
        }


        val allBookmarkList: List<BookmarkModel>
            get() = Select().from(BookmarkModel::class.java).execute()
    }
}